lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/DefaultNetworkServiceImpl.java

```diff
@@ -0,0 +1,217 @@
+      @Parameter(NetworkServiceParameters.NetworkServicePort.class) final int nsPort,
+      final TransportFactory transportFactory,
+      final NameResolver nameResolver) {
+
+    this.idFactory = idFactory;
+    this.connectionFactoryMap = new ConcurrentHashMap<>();
+    this.isStreamingCodecMap = new ConcurrentHashMap<>();
+    this.nsCodec = new DefaultNSMessageCodec(idFactory, connectionFactoryMap, isStreamingCodecMap);
+    this.nsLinkListener = new NetworkServiceLinkListener(connectionFactoryMap);
+    this.recvHandler = new NetworkServiceReceiveHandler(connectionFactoryMap, nsCodec);
+    this.nameResolver = nameResolver;
+    this.transport = transportFactory.newInstance(nsPort, recvHandler, recvHandler, new DefaultNSExceptionHandler());
+
+    this.nameServiceRegisteringStage = new SingleThreadStage<>(
+        "NameServiceRegisterer", new EventHandler<Tuple<Identifier, InetSocketAddress>>() {
+      @Override
+      public void onNext(final Tuple<Identifier, InetSocketAddress> tuple) {
+        try {
+          nameResolver.register(tuple.getKey(), tuple.getValue());
+          LOG.log(Level.FINEST, "Registered {0} with nameservice", tuple.getKey());
+        } catch (final Exception ex) {
+          final String msg = "Unable to register " + tuple.getKey() + " with name service";
+          LOG.log(Level.WARNING, msg, ex);
+          throw new RuntimeException(msg, ex);
+        }
+      }
+    }, 5);
+
+    this.nameServiceUnregisteringStage = new SingleThreadStage<>(
+        "NameServiceRegisterer", new EventHandler<Identifier>() {
+      @Override
+      public void onNext(final Identifier id) {
+        try {
+          nameResolver.unregister(id);
+          LOG.log(Level.FINEST, "Unregistered {0} with nameservice", id);
+        } catch (final Exception ex) {
+          final String msg = "Unable to unregister " + id + " with name service";
+          LOG.log(Level.WARNING, msg, ex);
+          throw new RuntimeException(msg, ex);
+        }
+      }
+    }, 5);
+  }
+
+  @Override
+  public synchronized <T> void registerConnectionFactory(final Class<? extends Name<String>> connectionFactoryId,
+                                                         final Codec<T> codec,
+                                                         final EventHandler<Message<T>> eventHandler,
+                                                         final LinkListener<Message<T>> linkListener) throws NetworkException {
+
+    final String connFactoryId = connectionFactoryId.getName();
+    if (connectionFactoryMap.get(connFactoryId) != null) {
+      throw new NetworkException("ConnectionFactory " + connectionFactoryId + " was already registered.");
+    }
+
+    connectionFactoryMap.put(connFactoryId, new NSConnectionFactory<>(this, connFactoryId, codec, eventHandler, linkListener));
+    isStreamingCodecMap.put(connFactoryId, codec instanceof StreamingCodec);
```

Comment:
It looks like we don't need to maintain this map separately. We can get the info from connectionFactoryMap.